src/util/treeJsonProcessing.ts

```diff
@@ -1,27 +1,29 @@
 import { getDefaultTreeState } from "../reducers/tree";
 import { getVaccineFromNode, getTraitFromNode, getDivFromNode } from "./treeMiscHelpers";
 import { calcFullTipCounts } from "./treeCountingHelpers";
+import { ReduxNode } from "../components/tree/phyloTree/types";
+import { TreeState } from "../components/tree/tree";
 
 const pseudoRandomName = () => (Math.random()*1e32).toString(36).slice(0, 6);
 
 /**
- * Adds certain properties to the nodes array - for each node in nodes it adds
- * node.fullTipCount - see calcFullTipCounts() description
- * node.hasChildren {bool}
- * node.arrayIdx  {integer} - the index of the node in the nodes array
- * @param  {array} nodes redux tree nodes
- * @return {Object} ret
- * @return {Set} ret.nodeAttrKeys collection of all `node_attr` keys whose values are Objects
- * @return {Array} ret.nodes input array (kinda unnecessary)
- *
- * side-effects: node.hasChildren (bool) and node.arrayIdx (INT) for each node in nodes
+ * Adds the following properties to each node:
+ * - fullTipCount
+ * - hasChildren
+ * - arrayIdx
```

Comment:
Yeah, there's an abundance of optional properties right now. It would be good to narrow those down by encoding the conditions somehow. I initially thought about putting those off in this PR by turning off `strictNullChecks`, but I think it's worth addressing now since it'll impact overall organization of types.

> Define `fullTipCount` as a non-optional property on `ReduxNode`

This is not trivial. The non-optional property will need to be used everywhere ReduxNode is initialized, such as: https://github.com/nextstrain/auspice/blob/5fccbc7d0aefd510e6aa68e162a699100a228859/src/util/treeJsonProcessing.ts#L74-L79

which will cause further conflicts such as the last line there where `ReduxNode.parent` is typed as `ReduxNode` which would require `fullTipCount` and error.

> `processNodes` would take an argument `nodes: Omit<ReduxNode, "fullTipCount">[]` and return `nodes: ReduxNode[]`

This wouldn't make any meaningful difference to the TypeScript compiler nor us. The docstring of this function makes it clear that `fullTipCount` is added here, and nothing is after calling this function to assert that `fullTipCount` is indeed set.

I think the only options are (note: I'm still kinda new to TypeScript)

1. non-null assertion operators (not great, [not allowed](https://typescript-eslint.io/rules/no-non-null-assertion) by `@typescript-eslint/recommended`)
2. type narrowing by handling alternate code paths where the value is unset

Reply:
> This is not trivial. The non-optional property will need to be used everywhere ReduxNode is initialized

Oh yeah, the nested nature of these objects makes changing types as data flows through functions difficult. I also don't have enough experience to know how to do this effectively in practice. Let's continue thinking about these things post this PR tho.

> This wouldn't make any meaningful difference to the TypeScript compiler nor us. The docstring of this function makes it clear that fullTipCount is added here, and nothing is done after calling this function to assert that fullTipCount is indeed set.

But it should make a big difference to something that uses a `ReduxNode` object and wants to use `fullTipCount` - currently tsc is always going to say "this property may be undefined" whereas ideally we'd have a typesystem which guaranteed it's presence or absence.

Reply:
@genehack and I discussed this general topic and one avenue we could explore (not in this PR!) is to have a type `A` with lots of optional properties that we can use as we build up the data structure then turn this into a type `B` once it arrives in redux and we know all/most properties have been defined. [Type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) may be useful here. Of everything in Auspice tree nodes are probably the worst to type due to the way we modify them in so many different places and do it on-the-fly rather than using immutable structures that redux suggests.

Reply:
I'm not quite sure if mapped types is the right approach for the `fullTipCount` example, but it seems related and I'll consider this in the same post-merge task

> - Post-merge: consider using mapped types in a new PR

Reply:
> turn this into a type `B` once it arrives in redux and we know all/most properties have been defined

This part isn't clear to me especially with the nested nature mentioned above. I'll create a separate issue to discuss further after this PR is merged.